fontir/src/ir.rs

```diff
@@ -991,14 +991,82 @@ impl GlyphAnchors {
             anchors,
         })
     }
+
+    /// `true` if any of our anchors are mark anchors
+    pub fn contains_marks(&self) -> bool {
+        self.anchors
+            .iter()
+            .any(|a| matches!(a.type_, AnchorType::Mark(_)))
+    }
+}
+
+/// The name shared by a group of anchors
+///
+/// This is used to determine the relationship between anchors: mark anchors
+/// in the group 'top' attach to the 'top' anchor of base glyphs.
+pub type GroupName = SmolStr;
+
+/// The type of an anchor.
+///
+/// Anchors serve multiple purposes in font sources. By convention, these different
+/// types of anchors are differentiated via naming convention. In IR, we parse
+/// these names to determine the intended type.
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
+pub enum AnchorType {
+    /// An attachment anchor on a base glyph
+    Base(GroupName),
+    /// An attachment anchor on a mark glyph
+    Mark(GroupName),
+    /// A base attachment on a ligature glyph
+    Ligature { group_name: GroupName, index: usize },
+    /// The enter anchor on a cursive glyph
+    //TODO: flesh these out when we use them. At least in glyphs there can
+    //be multiple (numbered) enter/exit anchors
+    Enter,
+    /// The exit anchor on a cursive gylph
+    Exit,
+    //TODO others? caret? vcaret?
+}
+
+impl AnchorType {
+    // this logic from
+    // <https://github.com/googlefonts/ufo2ft/blob/6787e37e6/Lib/ufo2ft/featureWriters/markFeatureWriter.py#L101>
+    pub(crate) fn new(name: impl AsRef<str>) -> Result<AnchorType, BadAnchorReason> {
+        let name = name.as_ref();
+        // _ prefix means mark. This convention appears to come from FontLab and is now everywhere.
+        if let Some(mark) = name.strip_prefix('_') {
+            if mark.is_empty() {
+                return Err(BadAnchorReason::NilMarkGroup);
+            }
+            if let Some((_, suffix)) = mark.rsplit_once('_') {
+                if suffix.parse::<usize>().is_ok() {
+                    return Err(BadAnchorReason::NumberedMarkAnchor);
+                }
+            }
+            return Ok(AnchorType::Mark(mark.into()));
+        } else if let Some((name, suffix)) = name.rsplit_once('_') {
+            // _1 suffix means a base in a ligature glyph
+            if let Ok(index) = suffix.parse::<usize>() {
+                if index == 0 {
+                    return Err(BadAnchorReason::ZeroIndex);
+                } else {
+                    return Ok(AnchorType::Ligature {
+                        group_name: name.into(),
+                        index,
+                    });
+                }
+            }
+        }
+        Ok(AnchorType::Base(name.into()))
+    }
 }
 
 /// A variable definition of an anchor.
 ///
 /// Must have at least one definition, at the default location.
 #[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
 pub struct Anchor {
-    pub name: AnchorName,
+    pub type_: AnchorType,
```

Comment:
how about `kind` to avoid the reserved `type` name and the need for `_`?

Reply:
good call, done :)